src/sentry/middleware/integrations/tasks.py

```diff
@@ -11,11 +13,100 @@
 from sentry.silo.base import SiloMode
 from sentry.silo.client import RegionSiloClient
 from sentry.tasks.base import instrumented_task
-from sentry.types.region import get_region_by_name
+from sentry.types.region import Region, get_region_by_name
 
 logger = logging.getLogger(__name__)
 
 
+@dataclass(frozen=True)
+class _AsyncResult:
+    region: Region
+    response: Response
+
+    def was_successful(self) -> bool:
+        return 200 <= self.response.status_code < 300
+
+
+@dataclass(frozen=True)
+class _AsyncRegionDispatcher(ABC):
+    request_payload: dict[str, Any]
+    response_url: str
+
+    @property
+    @abstractmethod
+    def log_code(self) -> str:
+        raise NotImplementedError
+
+    def log_message(self, tag: str) -> str:
+        return f"{self.log_code}.{tag}"
+
+    def dispatch(self, region_names: Iterable[str]) -> Response | None:
+        results = [self._dispatch_to_region(name) for name in region_names]
+        successes = [r for r in results if r.was_successful()]
+
+        logger.info(
+            self.log_message("async_region_response"),
+            extra={
+                "regions": [r.region.name for r in successes],
+                "response_map": {r.region.name: r.response.status_code for r in results},
+            },
+        )
+
+        if successes:
+            return self._forward_response(successes[-1])
```

Comment:
Using the last successful response replicates what the old code was doing in [lines 46-48](https://github.com/getsentry/sentry/pull/73998/files#diff-b2cbbebe9bbbe1cbfe72cb721b6efdd7c8fa211e0df2da77b39a5ecbccaf7947L46-L48). I'm not sure whether the original intent was

1. that we expect at most one successful response,
2. that we can forward any arbitrary one if we get more than one, or
3. that we intentionally should forward the last one.

I don't think it's 3 (because that would imply that the `region_names` argument is meaningfully ordered, which is not consistent with how this task is invoked AFAIK), but it was the easiest way to replicate the exact behavior. At least now it's a bit more explicit (`[-1]`) and we have a good place if we want to refine it.

Reply:
I think it is a combination of 1 & 2.

Reply:
Thanks. [Commented](https://github.com/getsentry/sentry/pull/73998/commits/095fc9a7d039dd4ac0ebc15abcc1002143087d00) accordingly.